### PR TITLE
refactoring

### PR DESCRIPTION
**Pull request by theahmedomar**
@vfdev-5 this is not complete yet , but I'd like your opinion on the current changes. , I think it's not testable on Colab due to huge computational requirements.

### CODE REVIEW

**Comment by vfdev-5** [comment]
@ahmedo42 thanks for the update ! I'll take a look a bit later.

File: train.py [hunk 1]
```diff
@@ -166,23 +185,131 @@ def update(engine, batch):
...
+    metrics = {
+        "Accuracy": Accuracy(output_transform=lambda x: (x[0][1], x[1][1])),
+        "nll": Loss(
+            torch.nn.CrossEntropyLoss(ignore_index=-100),
+            output_transform=lambda x: (x[0][0], x[1][0]),
+        ),
+    }
+    metrics.update(
+        {
+            "average_nll": RunningAverage(metrics["nll"]),
```
Comment (vfdev-5) [inline train.py]: Why do you use `RunningAverage` for validation ? Please, take a closer look at the original code, they `average_nll` to average the value accross distributed processes.

File: train.py [hunk 2]
```diff
@@ -192,84 +319,144 @@ def inference(engine, batch):
...
-                    "average_accuracy": MetricsLambda(average_distributed_scalar, metrics["accuracy"], args)})
-    metrics["average_ppl"] = MetricsLambda(math.exp, metrics["average_nll"])
-    for name, metric in metrics.items():
-        metric.attach(evaluator, name)
+    if args.dist_backend is not None:
+        trainer.add_event_handler(
+            Events.EPOCH_STARTED,
+            lambda engine: train_sampler.set_epoch(engine.state.epoch),
+        )
+        evaluator.add_event_handler(
```
Comment (vfdev-5) [inline train.py]: I think this is useless as there is no shuffling for validation dataloader... But they do that in their code. Let's keep in mind and remove it in another PR.

**Review by vfdev-5 — COMMENTED** [review]
Thanks for the update @ahmedo42 
I think we should revert code formatting in this PR and propose it in another one as it is very difficult to review 2 things: code-formatting and code improvements in the same PR.

**Comment by theahmedomar** [comment]
Thanks for the feedback , Now the code works on Colab without any issues , I suspect there is something wrong with progress bars as they are not displayed correctly
